### ISSUE TITLE
Training freezes during validation prediction

### ISSUE THREAD

**Issue by Xtaltec** [MIC-DKFZ/nnUNet#902]
Hi!

I'm training the network with the following configuration `nnUNet_train 3d_fullres nnUNetTrainerV2 TaskXXX_X all` and it successfully completes all epochs but then it freezes when predicting the validation set with no error messages. I have debugged the code and it seems that the code gets stuck in line 590 of nnUNetTrainer.py `export_pool = Pool(default_num_threads)`. I tested several of the solutions proposed in other issues such as reducing the RAM requirements by using `nnUNet_n_proc_DA=1` , but to no avail so far.  I know it is most likely related to the multiprocessing library, which is notoriously problematic, but I have not been able to find a solution.

Kind regards,
Xabier

**Comment by FabianIsensee** [comment]
Hi Xabier, unfortunately you will not find your luck with `nnUNet_n_proc_DA=1` in this case. What I suggest you do instead is modify the validation function such that it doesn't use multiprocesing for segmentation export.

Not fully working code but it should give you an idea:

change:
```
                results.append(export_pool.starmap_async(save_segmentation_nifti_from_softmax,
                                                         ((softmax_pred, join(output_folder, fname + ".nii.gz"),
                                                           properties, interpolation_order, self.regions_class_order,
                                                           None, None,
                                                           softmax_fname, None, force_separate_z,
                                                           interpolation_order_z),
                                                          )
                                                         )
                               )
```

into
```
                save_segmentation_nifti_from_softmax(softmax_pred, join(output_folder, fname + ".nii.gz"),
                                                           properties, interpolation_order, self.regions_class_order,
                                                           None, None,
                                                           softmax_fname, None, force_separate_z,
                                                           interpolation_order_z)
```

and remove the pool from the validate member function. validate is located in `nnUNetTrainer`.

You can also try `nnUNet_def_n_proc=1` first but I think this will only give problems.

Best,
Fabian

**Comment by Xtaltec** [comment]
Now it works perfectly, thanks.
